File: data-science-major/labs/course-6-r/python/11_dates.py
"""Experiment 11 (Python equivalent) -- working with dates.

R version: ../11_dates.R  (as.Date, format, lubridate)
"""
from datetime import date, datetime, timedelta

if __name__ == "__main__":
    print("PARSING                        R: as.Date('2026-08-26')")
    d = date.fromisoformat("2026-08-26")
    print(f"  ISO         '2026-08-26' -> {d}")
    d2 = datetime.strptime("26/08/2026", "%d/%m/%Y").date()
    print(f"  DD/MM/YYYY  '26/08/2026' -> {d2}   R: format='%d/%m/%Y'")
    print(f"  same date: {d == d2}")

    print("\nCOMPONENTS                     R: lubridate::year(), month(), day()")
    print(f"  year={d.year}  month={d.month}  day={d.day}")
    print(f"  weekday       = {d.strftime('%A')}   R: wday(d, label=TRUE)")
    print(f"  day of year   = {d.timetuple().tm_yday}")
    print(f"  ISO week      = {d.isocalendar().week}")

    print("\nARITHMETIC                     R: d + 30 ; difftime()")
    print(f"  d + 30 days   = {d + timedelta(days=30)}")
    print(f"  d - 7 days    = {d - timedelta(days=7)}")
    later = date.fromisoformat("2026-12-25")
    print(f"  days to {later} = {(later - d).days}")

    print("\nFORMATTING                     R: format(d, '%d %B %Y')")
    for fmt, label in (("%d-%m-%Y", "DD-MM-YYYY"), ("%d %B %Y", "long"),
                       ("%b %d, %Y", "abbreviated"), ("%Y-%m-%d", "ISO")):
        print(f"  {label:<12} {d.strftime(fmt)}")

    print("\nWHY DATES MUST NOT BE STRINGS")
    as_text = ["10/01/2026", "02/01/2026", "21/12/2025"]
    as_dates = [datetime.strptime(x, "%d/%m/%Y").date() for x in as_text]
    print(f"  sorted as text : {sorted(as_text)}")
    print(f"  sorted as dates: {[str(x) for x in sorted(as_dates)]}")
    print("  Text sorting puts 02/01 before 10/01 before 21/12 -- alphabetical,")
    print("  not chronological. The 2025 date ends up LAST. This is exactly the")
    print("  bug that makes date columns stored as VARCHAR dangerous (Course 5).")

    assert sorted(as_dates)[0].year == 2025, "chronological order must start in 2025"
    assert sorted(as_text)[0].startswith("02"), "text order must start with 02"
    print("\n  demonstrated: text order and date order genuinely differ ✓")

File: data-science-major/labs/course-6-r/11_dates.R
# =====================================================================
# NOT EXECUTED IN VERIFICATION -- R is not installable in this environment
# (Debian repos blocked). Desk-checked only; numbers in comments come from
# the executed Python equivalent in python/. Run this in RStudio.
# =====================================================================
# Experiment 11: Working with dates and times
# Python equivalent: python/11_dates.py

# --- BASE R ---
d <- as.Date("2026-08-26")                       # ISO is the safe default
d2 <- as.Date("26/08/2026", format = "%d/%m/%Y")
Sys.Date(); Sys.time()

format(d, "%d-%m-%Y")      # "26-08-2026"
format(d, "%d %B %Y")      # "26 August 2026"
weekdays(d); months(d)

d + 30                     # date arithmetic works directly
difftime(as.Date("2026-12-25"), d, units = "days")   # 121 days

# --- lubridate: much easier ---
library(lubridate)
ymd("2026-08-26"); dmy("26-08-2026"); mdy("08-26-2026")
year(d); month(d); day(d); wday(d, label = TRUE)
d + days(30); d + months(1); d + years(1)
d %m+% months(1)     # SAFE month addition
# 31 Jan %m+% months(1) gives 28 Feb, not an invalid 31 Feb.

# --- WHY DATES MUST NOT BE STORED AS TEXT ---
as_text <- c("10/01/2026", "02/01/2026", "21/12/2025")
sort(as_text)
#   "02/01/2026" "10/01/2026" "21/12/2025"
#   ALPHABETICAL: the 2025 date sorts LAST. This is wrong and silent.

as_dates <- dmy(as_text)
sort(as_dates)
#   "2025-12-21" "2026-01-02" "2026-01-10"   <- correct chronological order

# Same lesson as Course 5: a date column stored as VARCHAR sorts and compares
# alphabetically, which is almost never what you want.

# --- FORMAT CODES (also used by format() in base R) ---
#   %Y 4-digit year   %y 2-digit year   %m month number
#   %B full month     %b abbreviated    %d day of month
#   %A full weekday   %a abbreviated    %H:%M:%S time
